### PR TITLE
Also compile libcore.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by chriseth** [comment]
Ignored tests pass locally.

File: src/riscv/parser.rs [hunk 1]
```diff
@@ -98,33 +101,50 @@ pub fn parse_asm(input: &str) -> Vec<Statement> {
...
+
+pub fn ends_control_flow(s: &Statement) -> bool {
+    match s {
+        Statement::Instruction(instruction, _) => match instruction.as_str() {
+            "li" | "lui" | "mv" | "add" | "addi" | "sub" | "neg" | "mul" | "mulhu" | "xor"
+            | "xori" | "and" | "andi" | "or" | "ori" | "not" | "slli" | "sll" | "srli" | "srl"
+            | "seqz" | "snez" | "slti" | "sltu" | "sltiu" | "beq" | "beqz" | "bgeu" | "bltu"
+            | "blt" | "bge" | "bltz" | "blez" | "bgtz" | "bgez" | "bne" | "bnez" | "jal"
+            | "jalr" | "call" | "ecall" | "lw" | "lb" | "lbu" | "sw" | "sh" | "sb" => false,
+            "j" | "jr" | "tail" | "ret" | "unimp" => true,
```
Comment (leonardoalt) [inline src/riscv/parser.rs:141]: Why are `jal` and `jalr` not control-flow-ending?

File: src/riscv/parser.rs [hunk 2]
```diff
@@ -136,11 +156,30 @@ pub fn extract_data_objects(statements: &[Statement]) -> BTreeMap<String, Vec<u8
                 current_label = Some(l.as_str());
             }
             // TODO We ignore size and alignment directives.
+            // TODO this might all be totally wrong
```
Comment (leonardoalt) [inline src/riscv/parser.rs:159]: lol how do we figure it out

File: src/riscv/parser.rs [hunk 3]
```diff
@@ -136,11 +156,30 @@ pub fn extract_data_objects(statements: &[Statement]) -> BTreeMap<String, Vec<u8
...
-                    objects.insert(name.clone(), None);
+                    // TODO do we need to consider name clashes?
+                    if !objects.contains_key(name) {
+                        objects.insert(name.clone(), None);
+                    }
+                }
+                (
+                    ".zero",
+                    [Argument::Constant(Constant::Number(n))]
+                    // TODO not clear what the second argument is
```
Comment (leonardoalt) [inline src/riscv/parser.rs:172]: Maybe it's the number of 0 bytes?

Comment (leonardoalt) [reply src/riscv/parser.rs:172]: Ah the 2nd argument, weird, the spec I'm following only has one argument.

File: src/riscv/compiler.rs [hunk 4]
```diff
@@ -43,6 +29,62 @@ pub fn compile_riscv_asm(data: &str) -> String {
     output
 }
 
+fn filter_reachable_from(
+    label: &str,
+    statements: &mut Vec<Statement>,
+    objects: &mut BTreeMap<String, Vec<u8>>,
+) {
```
Comment (leonardoalt) [inline src/riscv/compiler.rs:36]: Why not take statements and objects by value and return by value?

File: src/riscv/compiler.rs [hunk 5]
```diff
@@ -43,6 +29,62 @@ pub fn compile_riscv_asm(data: &str) -> String {
...
+    let label_offsets = parser::extract_label_offsets(statements);
+    let mut code: Vec<Statement> = vec![];
+    let mut queued_labels: BTreeSet<&str> = vec![label].into_iter().collect();
```
Comment (leonardoalt) [inline src/riscv/compiler.rs:39]:
```suggestion
    let mut queued_labels: BTreeSet<&str> = vec![label].into();
```
Doesn't this work?

Comment (leonardoalt) [inline src/riscv/compiler.rs:39]:
```suggestion
    let mut queued_labels = BTreeSet<&str>::from(vec![label]);
```
or this

File: src/riscv/compiler.rs [hunk 6]
```diff
@@ -43,6 +29,62 @@ pub fn compile_riscv_asm(data: &str) -> String {
...
+    let mut referenced_labels: BTreeSet<&str> = vec![label].into_iter().collect();
```
Comment (leonardoalt) [inline src/riscv/compiler.rs:40]: Same here

File: src/riscv/compiler.rs [hunk 7]
```diff
@@ -183,6 +231,8 @@ instr mload -> X { { addr, STEP, X } is m_is_read { m_addr, m_step, m_value } }
 // ============== control-flow instructions ==============
 
 instr jump l: label { pc' = l }
+instr jump_dyn X { pc' = X }
+instr jump_and_link_dyn X { pc' = X, x1' = pc + 1 }
```
Comment (leonardoalt) [inline src/riscv/compiler.rs:235]: why `x1'`?

File: src/riscv/compiler.rs [hunk 8]
```diff
@@ -197,10 +247,17 @@ instr branch_if_positive X, l: label {
     X + 2**32 - 1 = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 + wrap_bit * 2**32,
     pc' = wrap_bit * l + (1 - wrap_bit) * (pc + 1)
 }
+// input X is required to be the difference of two 32-bit unsigend values.
+// i.e. -2**32 < X < 2**32
+instr is_positive X -> Y {
+    X + 2**32 - 1 = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 + wrap_bit * 2**32,
+    Y = wrap_bit
+}
```
Comment (leonardoalt) [inline src/riscv/compiler.rs:255]: If `X` is out of the pre-condition range, can't `wrap_bit` be non-binary?

File: src/riscv/compiler.rs [hunk 9]
```diff
@@ -369,127 +426,94 @@ pil{
...
-}
-*/
-            r#"
-	li	a3, 4
-	blt	a2, a3, __memset_LBB5_5
-	li	a5, 0
-	lui	a3, 4112
-	addi	a3, a3, 257
-	mul	a6, a1, a3
+    */
```
Comment (leonardoalt) [inline src/riscv/compiler.rs:474]: Is this commented stuff gonna change in this PR?

File: src/riscv/compiler.rs [hunk 10]
```diff
@@ -43,6 +29,62 @@ pub fn compile_riscv_asm(data: &str) -> String {
...
+            referenced_labels.extend(parser::referenced_labels(s))
+        }
+        code.push(s.clone());
+        if parser::ends_control_flow(s) {
+            break;
+        }
+    }
+    (code, referenced_labels.into_iter().collect(), seen_labels)
+}
+
```
Comment (leonardoalt) [inline src/riscv/compiler.rs:87]: It would be nice to have some tests for the functions above...

File: src/riscv/parser.rs (hunk 1 above)
Comment (chriseth) [reply src/riscv/parser.rs:141]: It's a `call`, it stores the address of the following instruction in a register.

Comment (chriseth) [reply src/riscv/parser.rs:141]: `jal` is "jump and link"

File: src/riscv/compiler.rs (hunk 7 above)
Comment (chriseth) [reply src/riscv/compiler.rs:235]: It's the spec as far as I understand.

File: src/riscv/compiler.rs (hunk 8 above)
Comment (chriseth) [reply src/riscv/compiler.rs:255]: but wrap_bit is constrained to be binary somewhere else (unconditionally)

File: src/riscv/compiler.rs (hunk 9 above)
Comment (chriseth) [reply src/riscv/compiler.rs:474]: No, it's unchanged. We should directly compile from rust at some point.

**Comment by chriseth** [comment]
Replaced by https://github.com/chriseth/powdr/pull/187

**Pull request closed without merging**